File: backend/app/schemas/replay.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel


class ReplayPlayerInfo(BaseModel):
    display_name: str
    is_bot: bool


class ReplayListItem(BaseModel):
    index: int
    game_id: UUID
    display_label: str
    human_player_names: List[str]
    played_date: str
    created_at: datetime
    num_players: int
    winner: Optional[str] = None
    players: List[ReplayPlayerInfo]


class ReplayListResponse(BaseModel):
    replays: List[ReplayListItem]
    page: int
    size: int
    total_items: int
    total_pages: int


class ReplayFrame(BaseModel):
    frame_index: int
    step: Optional[int] = None
    action: Optional[str] = None
    commentary: Optional[str] = None
    rich_state: Dict[str, Any]


class ReplayDetailResponse(BaseModel):
    game_id: UUID
    display_label: str
    players: List[ReplayPlayerInfo]
    replay_frames: List[ReplayFrame]
    total_frames: int
    final_scores: List[Dict[str, Any]] = []

File: frontend/src/types/replay.ts
import type { GameState } from './gameState';

export interface ReplayPlayerInfo {
  display_name: string;
  is_bot: boolean;
}

export interface ReplayListItem {
  index: number;
  game_id: string;
  display_label: string;
  human_player_names: string[];
  played_date: string;
  created_at: string;
  num_players: number;
  winner: string | null;
  players: ReplayPlayerInfo[];
}

export interface ReplayListResponse {
  replays: ReplayListItem[];
  page: number;
  size: number;
  total_items: number;
  total_pages: number;
}

export interface ReplayFrame {
  frame_index: number;
  step: number | null;
  action: string | null;
  commentary: string | null;
  rich_state: GameState;
}

export interface ReplayDetailResponse {
  game_id: string;
  display_label: string;
  players: ReplayPlayerInfo[];
  replay_frames: ReplayFrame[];
  total_frames: number;
  final_scores: Array<Record<string, unknown>>;
}
